ssl/tls13_enc.c

```diff
@@ -667,7 +676,7 @@ int tls13_export_keying_material(SSL *s, unsigned char *out, size_t olen,
     unsigned int hashsize, datalen;
     int ret = 0;
 
-    if (ctx == NULL || !SSL_is_init_finished(s))
```

Comment:
I kind of think there should be some test here to prevent you from exporting keying material too early.

Reply:
Do you have any idea about this?  I thought that it requires the knowledge of openssl internals about session state.

Reply:
Some ideas:

1) Add a new flag somewhere to the SSL structure to indicate that export key material is available. This is simple and should always work, but has the disadvantage of growing the size of the SSL struct.

2) Perhaps check the value of `s->s3->previous_server_finished_len` is non-zero? This gets set during the processing or construction of the server Finished message. Technically there is a small window between construction of the server Finished and the subsequent change of cipher state where the exporter master secret is generated where control could return to the user application. If the application attempts to export keying material in that window then it will get the wrong result. This is probably the simplest solution but isn't perfect

3) Building on option 2 we could additionally check the value of `s->statem.hand_state` to check that it isn't `TLS_ST_SW_FINISHED` (if it is then it could indicate we are in that small window). I prefer to keep all accesses of the state machine internal state within the state machine code itself and not littered around the rest of the code. Therefore I'd probably create a function to check this in ssl/statem/statem.c, in much the same way as the `ossl_statem_skip_early_data` function works now.

On reflection option 3, while a bit more complicated, is probably the way to go.

Reply:
Thank you for the pointer.  I'll go with option 3.

Reply:
I added ossl_statem_export_allowed in ssl/statem/statem.{c,h}, but build fails:

```
gcc -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/home/foo/lib/engines-1.1\"" -Wall -O3 -pthread -m64 -DL_ENDIAN     -o test/tls13secretstest ssl/packet.o ssl/tls13_enc.o test/tls13secretstest.o \
	  -L. -lssl test/libtestutil.a -L. -lcrypto -ldl 
ssl/tls13_enc.o: In function `tls13_export_keying_material':
tls13_enc.c:(.text+0x15f3): undefined reference to `ossl_statem_export_allowed'
collect2: error: ld returned 1 exit status
```

statem.h has the following declaration:

```c
__owur int ossl_statem_export_allowed(SSL *s);
```

It is strange that libssl.so.1.1 includes that symbol:

```
$ nm libssl.so.1.1|grep ossl_statem_export_allowed
000000000004d1f0 t ossl_statem_export_allowed
```

Do you have any idea how to fix it?

Reply:
Ah, so it is internal symbol, not exported, and linking with shared lib fails.

Reply:
tls13secretstest is slightly unusual. You need to add a mocked out version of the symbol. Just something that always returns 1 should do it. See the section of tls13secretstest.c that starts with the comment "Mocked out implementations of various functions" and ends with "End of mocked out code"

Reply:
It works.  Thank you!

Reply:
be25f3b does options 3.